File: php/ambler.php
<?php

class Next {
    private $nextFunc;
    public mixed $state;

    public function __construct(callable $nextFunc, mixed $state) {
        $this->nextFunc = $nextFunc;
        $this->state = $state;
    }

    public function run(): ?Next {
        return ($this->nextFunc)($this->state);
    }
}

function node(callable $factory): callable {
    return fn($state) => $factory()($state);
}

function amble(callable $initial, mixed $state): void {
    $next = $initial($state);
    while ($next !== null) {
        $next = $next->run();
    }
}

File: ruby/lib/ambler.rb
# frozen_string_literal: true

require 'async'

module Ambler
  class Next
    attr_reader :state

    def initialize(next_func, state)
      @next_func = next_func
      @state = state
    end

    def run
      @next_func.call(@state)
    end
  end

  def self.node(&factory)
    ->(state) { factory.call.call(state) }
  end

  def self.amble(initial, state)
    Async do
      next_val = initial.call(state)
      next_val = next_val.run while next_val
    end
  end
end
